### DISCUSSION TITLE
Lock field editing by keyword in field

### DISCUSSION THREAD

**Discussion by jbk68** [cortexrd/Knack-Toolkit-Library#382] [General]
Not sure if KTL currently or has the potential of doing the following:

I have a particular field (Notes) that I have allowed the end user to add or edit notes to.  What I would like to do is, if I edit or add a note to this field and have a particular keyword in the notes, this would disable anyone but an Admin to edit or override notes.

**Here's an example:**

Note field contains "move equipment to Dallas location".  In this case, the word "move" would be the magic keyword which would disable anyone WITHOUT admin rights to edit (delete/modify etc...)

**Another way that this could be done is:**

I have a multiple choice field that is only editable by Admin's.  If this multiple choice was set to "MOVE", this would disable the ability for anyone without Admin rights to edit Notes field until the multiple choice field was set to something other than "MOVE"

Any and all help would be appreciated... JON

**Comment by Deanobeano2** [comment]
Hi Jon

Without researching this in detail, couldn't this be achieved natively in Knack?

My thought would be to create a Yes/No field for "Is note editable" with the default value being Yes. Then use a Record Action to Update this record to field value = No when your criteria is met.

Dean

**Comment by jbk68** [comment]
Hi Dean, that you for the response, but I've tried something similar to your suggestion natively in Knack, but can't seem to make it work as expected.  

Any and all suggestions including a script would be highly appreciated... JON

**Comment by Deanobeano2** [comment]
Hi Jon

Here's a "proof of concept" I quickly knocked up by manually setting the "Is note editable" values.

![image](https://github.com/user-attachments/assets/c6fa97e9-8537-464a-a081-2e5b1ece1a6b)

![image](https://github.com/user-attachments/assets/ae13b9ec-2cd7-4d6c-a455-9d47905be690)

Glad to have gone through this quick exercise because until now I didn't realize you could set display rules for the buttons.

Dean

**Comment by jbk68** [comment]
Hi Dean, I thank you for your "proof of concept" but I'm not sure how this will achieve what I asked in my initial post.  I'm looking for a way to "disable" the ability for any one without  "admin" rights the ability to edit a current note if that note (field) contains a certain keyword.

In my initial post, as an example, I used the keyword "move" to make the notes field "locked" to anyone without admin rights.

Example: notes field is "**move** equipment to Dallas".  Notes field is now **NOT** editable, except by admin because of keyword "move"
Example: notes field is "equipment located in Dallas".  Notes field is editable by all users.

With the help of OpenGPT I tried several JavaScripts with no luck.  Seems like I'm fighting Knack's built in field editable settings in builder.  I tried multiple scripts and turn edit field off/on with no luck

Again thank you for your help and if you find a solution, please let me know... JON

**Reply by CSWinnall** [reply]
Hi Jon 

You can make an inline edit uneditable by removing the class cell-edit.

You can also use KTL function ktl.account.matchUserRoles(['Admin'])

And
$('td.field_123:contains("Move"')).removeClass('cell-edit)

That should get you started with GPT 

Craig

**Reply by jbk68** [reply]
Hi Craig, thanks for the response.  I'm pretty sure I tried something like your suggestion and couldn't get it to work.  I didn't try the KTL function YET, just trying to get the field to NOT be editable with keyword "MOVE".  Most likely something in my code was wrong.

Thanks again... JON

I

**Comment by Deanobeano2** [comment]
Craig

Thanks for stepping in with a solution.

Dean

**Comment by CSWinnall** [comment]
Can you post your code and I'll take a look in the morning?


Craig

**Comment by jbk68** [comment]
Yes.  I'm going to give it one more try with GPT and specifically use your
suggestions. Regardless if I get a working solution or not, I'll post the
script.

Thanks again Craig and Dean... JON

**Comment by jbk68** [comment]
Craig, finally got it to work using your suggestion.  Below is the script.  I even added a lock icon when user hovers over field.  Thanks for your help

$(document).on('knack-view-render.view_85', function(event, view) {
  // Target the table cell containing field_51 that contains the word "MOVE"
  $('td.field_51').filter(function() {
    return $(this).text().toUpperCase().includes("MOVE");
  }).each(function() {
    // Remove the cell-edit class to disable editing
    $(this).removeClass('cell-edit');
    
    // Append the lock icon but hide it initially, styled in black
    $(this).append(' <i class="fa fa-lock lock-icon" aria-hidden="true" style="display: none; color: black;"></i>');
  });
});

// Add hover effect to show the lock icon when hovering over the field_51 cell
$(document).on('mouseenter', 'td.field_51', function() {
  if ($(this).text().toUpperCase().includes("MOVE")) {
    $(this).find('.lock-icon').css('display', 'inline-block');
  }
}).on('mouseleave', 'td.field_51', function() {
  $(this).find('.lock-icon').css('display', 'none');
});

**Reply by CSWinnall** [reply]
That's Great Jon

This will do it for everyone but if you want admin users to be able to edit then you can add this (Replace "Admin" with the user role that is allowed to edit) :
```
    $(document).on('knack-view-render.view_85', function(event, view) {
        // Target the table cell containing field_51 that contains the word "MOVE"
        if (!ktl.account.matchUserRoles(['Admin'])) {
            $('td.field_51').filter(function () {
                return $(this).text().toUpperCase().includes("MOVE");
            }).each(function () {
                // Remove the cell-edit class to disable editing
                $(this).removeClass('cell-edit');
            
                // Append the lock icon but hide it initially, styled in black
                $(this).append(' <i class="fa fa-lock lock-icon" aria-hidden="true" style="display: none; color: black;"></i>');
            });
        }
    });
```

**Reply by jbk68** [reply]
Thank you...  

Maybe this is something that can be added to KTL where a field could be locked from being edited based on keyword ???
